### DISCUSSION TITLE
How to run multiple scala operations on one Spark engine?

### DISCUSSION THREAD

**Discussion by HaoYang670** [apache/kyuubi#6853] [Q&A] [answered]
We want to run mulitple scala operations (in different sessions) on one Spark engine. We find Kyuubi doesn't support it by default. Only one operation is in RUNNING state and others are pending. I don't know if there is a configuration for that or Kyuubi doesn't support it at all.

**Comment by HaoYang670** [comment]
I find a PR that add a static lock for all operations, https://github.com/apache/kyuubi/pull/4150. Can we remove that?

**Reply by turboFei** [reply, marked as answer]
> I find a PR that add a static lock for all operations, https://github.com/apache/kyuubi/pull/4150. Can we remove that?

No. we have to disable the concurrent scala code interpret. because they share the same spark.repl.class.outputDir, which is a static configuration and unable to change it once the Spark context is initialized. 

https://github.com/apache/kyuubi/blob/b265ccbe2d850ce63b568b13b37713ec198b1505/externals/kyuubi-spark-sql-engine/src/main/scala-2.12/org/apache/kyuubi/engine/spark/repl/KyuubiSparkILoop.scala#L49-L50

https://github.com/apache/spark/blob/332efb2eabb2b9383cfcfb0bf633089f38cdb398/core/src/main/scala/org/apache/spark/SparkContext.scala#L497-L500 

See the bug mentioned in https://github.com/apache/kyuubi/issues/3302

**Comment by pan3793** [comment]
Situations might be changed with SPARK-45856 and a series of changes related to Spark Connect, @HaoYang670 you are welcome to evaluate them and report the conclusion.
